proto/pulumi/provider.proto

```diff
@@ -363,81 +373,135 @@ message CheckFailure {
+    string name = 7;
+
+    // The type of the resource being diffed. This must match the type specified by the `urn` field, and is passed so
+    // that providers do not have to implement URN parsing in order to extract the type of the resource.
+    string type = 8;
 }
 
+// `PropertyDiff` describes the kind of change that occurred to a property during a diff operation. A `PropertyDiff` may
+// indicate that a property was added, deleted, or updated, and may further indicate that the change requires a
+// replacement.
 message PropertyDiff {
-    Kind kind = 1; // The kind of diff asdsociated with this property.
-    bool inputDiff = 2; // The difference is between old and new inputs, not old and new state.
+    // The kind of diff associated with this property.
+    Kind kind = 1;
+
+    // True if and only if this difference represents one between a pair of old and new inputs, as opposed to a pair of
+    // old and new states.
+    bool inputDiff = 2;
 
+    // The type of property diff kinds.
     enum Kind {
-        ADD = 0;            // this property was added
-        ADD_REPLACE = 1;    // this property was added, and this change requires a replace
-        DELETE = 2;         // this property was removed
-        DELETE_REPLACE = 3; // this property was removed, and this change requires a replace
-        UPDATE = 4;         // this property's value was changed
-        UPDATE_REPLACE = 5; // this property's value was changed, and this change requires a replace
+        // This property was added.
+        ADD = 0;
+
+        // This property was added, and this change requires a replace.
+        ADD_REPLACE = 1;
+
+        // This property was removed.
+        DELETE = 2;
+
+        // This property was removed, and this change requires a replace.
+        DELETE_REPLACE = 3;
+
+        // This property's value was changed.
+        UPDATE = 4;
+
+        // This property's value was changed, and this change requires a replace.
+        UPDATE_REPLACE = 5;
     }
 }
 
+// `DiffResponse` is the type of responses sent by a [](pulumirpc.ResourceProvider.DiffConfig) or
+// [](pulumirpc.ResourceProvider.Diff) call. A `DiffResponse` indicates whether a resource is unchanged, requires
+// updating (that is, can be changed "in place"), or requires replacement (that is, must be destroyed and recreated
+// anew). Legacy implementations may also signal that it is unknown whether there are changes or not.
+//
+// `DiffResponse` has evolved since its inception and there are now a number of ways that providers can signal their
+// intent to callers:
+//
+// * *Simple diffs* utilise the `changes` field to signal which fields are responsible for a change, and the `replaces`
+//   field to further communicate which changes (if any) require a replacement as opposed to an update.
+//
+// * *Detailed diffs* are those with `hasDetailedDiff` set, and utilise the `detailedDiff` field to provide a more
```

Comment:
Worth noting that providers not implementing detailed Diffs will have the engine compute them. Worth noting also that detailed diffs only (I hope this is still true?) affect how the changes are displayed to the user, not what is actually going to happen.

Reply:
Tracking in #17691